circuit/src/native.rs

```diff
@@ -0,0 +1,628 @@
+				let mut ops_i = self.ops.get(&pk_i).unwrap().clone();
+
+				// Give zero score for peers
+				//   - giving scores to itself
+				// 	 - giving scores to default peer
+				//
+				// Example:
+				//   Peer1 opinion
+				//		[(p1, 10), (p2, 10), (null, 10)]
+				// 	 => [(p1,  0), (p2, 10), (null,  0)]
+				//
+				for j in 0..NUM_NEIGHBOURS {
+					let (pk_j, _) = ops_i.scores[j].clone();
+					if pk_j == pk_i || pk_j == PublicKey::default() {
+						ops_i.scores[j] = (pk_j, Fr::zero());
+					};
+				}
+
+				// Update the opinion array - pairs of (key, score)
+				//
+				// Example 1:
+				// 	filtered_set => [p1, null, p3]
+				//	Peer1 opinion
+				// 		[(p1, 0), (p6, 10),  (p3, 10)]
+				//   => [(p1, 0), (null, 0), (p3, 10)]
+				//
+				// Example 2:
+				// 	filtered_set => [p1, p2, p3]
+				//	Peer1 opinion
+				// 		[(p1, 0), (p3, 10), (null, 0)]
+				//   => [(p1, 0), (p2, 0),  (p3, 0)]
+				for j in 0..NUM_NEIGHBOURS {
+					let (set_pk_j, _) = filtered_set[j];
+					let (op_pk_j, _) = ops_i.scores[j].clone();
+					if set_pk_j != op_pk_j {
+						ops_i.scores[j] = (set_pk_j, Fr::zero());
+					}
+				}
+
+				// Distribute the scores
+				//
+				// Example 1:
+				// 	filtered_set => [p1, p2, p3]
+				//	Peer1 opinion
+				// 		[(p1, 0), (p2, 0), (p3, 10)]
+				//   => [(p1, 0), (p2, 0), (p3, 10)]
+				//
+				// Example 2: (Cont.d from above example 2)
+				// 	filtered_set => [p1, p2, p3]
+				//	Peer1 opinion
+				//      [(p1, 0), (p2, 0), (p3, 0)]
+				//   => [(p1, 0), (p2, 1), (p3, 1)]
+				let op_score_sum =
+					ops_i.scores.iter().fold(Fr::zero(), |acc, &(_, score)| acc + score);
+				if op_score_sum == Fr::zero() {
+					for j in 0..NUM_NEIGHBOURS {
+						let (pk_j, _) = ops_i.scores[j].clone();
+						if pk_j != pk_i && pk_j != PublicKey::default() {
+							ops_i.scores[j] = (pk_j, Fr::from(1));
+						}
+					}
+				}
+
+				filtered_ops.insert(pk_i, ops_i);
+			}
+		}
+
+		(filtered_set, filtered_ops)
+	}
+}
+
+#[cfg(test)]
+mod test {
+	use super::{EigenTrustSet, Opinion, INITIAL_SCORE, NUM_NEIGHBOURS};
+	use crate::{
+		calculate_message_hash,
+		eddsa::native::{sign, PublicKey, SecretKey},
+	};
+	use halo2::halo2curves::{bn256::Fr, FieldExt};
+	use rand::thread_rng;
+
+	fn sign_opinion(
+		sk: &SecretKey, pk: &PublicKey, pks: &[PublicKey; NUM_NEIGHBOURS],
+		scores: &[Fr; NUM_NEIGHBOURS],
+	) -> Opinion {
+		let (_, message_hashes) =
+			calculate_message_hash::<NUM_NEIGHBOURS, 1>(pks.to_vec(), vec![scores.to_vec()]);
+		let sig = sign(sk, pk, message_hashes[0]);
+
+		let scores = pks.zip(*scores);
+		let op = Opinion::new(sig, message_hashes[0], scores);
+		op
+	}
+
+	#[test]
+	#[should_panic]
+	fn test_add_member_in_initial_set() {
+		let mut set = EigenTrustSet::new();
+
+		let rng = &mut thread_rng();
+
+		let sk1 = SecretKey::random(rng);
+		let pk1 = sk1.public();
+
+		set.add_member(pk1);
+
+		// Re-adding the member should panic
+		set.add_member(pk1);
+	}
+
+	#[test]
+	#[should_panic]
```

Comment:
I this we could also avoid panic in this scenario, by:

If member is in the set but didn't add his opinion, we should assign default one, e.g.:
if `set = [peer1, peer2, peer3]`
if `peer1` didn't add his opinion, we assign `peer1_op = [(peer1, 0), (peer2, 1), (peer3, 1)]`